tests/ApiTest.php

```diff
@@ -0,0 +1,3669 @@
+	{
+		$this->assertStringStartsWith('<?xml version="1.0"?>', $result);
+		$this->assertContains('<'.$root_element, $result);
+		// We could probably do more checks here.
+	}
+
+	/**
+	 * Get the path to a temporary empty PNG image.
+	 * @return string Path
+	 */
+	private function getTempImage()
+	{
+		$tmpFile = tempnam(sys_get_temp_dir(), 'tmp_file');
+		file_put_contents(
+			$tmpFile,
+			base64_decode(
+				// Empty 1x1 px PNG image
+				'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8/5+hHgAHggJ/PchI7wAAAABJRU5ErkJggg=='
+			)
+		);
+
+		return $tmpFile;
+	}
+
+	/**
+	 * Test the api_user() function.
+	 * @return void
+	 */
+	public function testApiUser()
+	{
+		$this->assertEquals($this->selfUser['id'], api_user());
+	}
+
+	/**
+	 * Test the api_user() function with an unallowed user.
+	 * @return void
+	 */
+	public function testApiUserWithUnallowedUser()
+	{
+		$_SESSION = ['allow_api' => false];
+		$this->assertEquals(false, api_user());
+	}
+
+	/**
+	 * Test the api_source() function.
+	 * @return void
+	 */
+	public function testApiSource()
+	{
+		$this->assertEquals('api', api_source());
+	}
+
+	/**
+	 * Test the api_source() function with a Twidere user agent.
+	 * @return void
+	 */
+	public function testApiSourceWithTwidere()
+	{
+		$_SERVER['HTTP_USER_AGENT'] = 'Twidere';
+		$this->assertEquals('Twidere', api_source());
+	}
+
+	/**
+	 * Test the api_source() function with a GET parameter.
+	 * @return void
+	 */
+	public function testApiSourceWithGet()
+	{
+		$_GET['source'] = 'source_name';
+		$this->assertEquals('source_name', api_source());
+	}
+
+	/**
+	 * Test the api_date() function.
+	 * @return void
+	 */
+	public function testApiDate()
+	{
+		$this->assertEquals('Wed Oct 10 00:00:00 +0000 1990', api_date('1990-10-10'));
+	}
+
+	/**
+	 * Test the api_register_func() function.
+	 * @return void
+	 */
+	public function testApiRegisterFunc()
+	{
+		global $API;
+		$this->assertNull(
+			api_register_func(
+				'api_path',
+				function () {
+				},
+				true,
+				'method'
+			)
+		);
+		$this->assertTrue($API['api_path']['auth']);
+		$this->assertEquals('method', $API['api_path']['method']);
+		$this->assertTrue(is_callable($API['api_path']['func']));
+	}
+
+	/**
+	 * Test the api_login() function without any login.
+	 * @return void
+	 * @runInSeparateProcess
+	 * @expectedException Friendica\Network\HTTPException\UnauthorizedException
+	 */
+	public function testApiLoginWithoutLogin()
+	{
+		api_login($this->app);
+	}
+
+	/**
+	 * Test the api_login() function with a bad login.
+	 * @return void
+	 * @runInSeparateProcess
+	 * @expectedException Friendica\Network\HTTPException\UnauthorizedException
+	 */
+	public function testApiLoginWithBadLogin()
+	{
+		$_SERVER['PHP_AUTH_USER'] = 'user@server';
+		api_login($this->app);
+	}
+
+	/**
+	 * Test the api_login() function with oAuth.
+	 * @return void
+	 */
+	public function testApiLoginWithOauth()
+	{
+		$this->markTestIncomplete('Can we test this easily?');
```

Comment:
Is there any way to check how other projects are testing for this? This would require an elaborate mockup, especially to test the callback URL.

Reply:
We could use a library like [Mockery](https://github.com/mockery/mockery) in order to mock the `FKOAuth1` class.

Reply:
Don't mind me, I'm just throwing around technobabble I'm hoping means something, but I've never written any test in my life.